src/diff/children.js

```diff
@@ -53,6 +54,13 @@ export function diffChildren(
 	isHydrating,
 	refQueue
 ) {
+	if (
+		newParentVNode.type === 'template' &&
+		excessDomChildren?.length === 0 &&
+		parentDom instanceof DocumentFragment
+	) {
```

Comment:
Let's make this fast by adding the `isHydrating` check, we don't need to check `excessDomChildren` length as we expect this to be 0 as per your assumptions in this PR. I would assume that we are in a DocumentFragment when we are dealing with a template.

```suggestion
	if (
	    isHydrating &&
		newParentVNode.type === 'template'
	) {
```

I assume you could save even more bytes by moving this to https://github.com/preactjs/preact/blob/1dae4898efd3c4a4e596917ee9426ed617efe50b/src/diff/index.js#L596-L611 and changing this condition to use `newVNode.type === 'template'` as well as hoisting the duplicated check